File: Source/SafetySharp/CSharp/Transformation/SymbolMap.cs
// The MIT License (MIT)
// 
// Copyright (c) 2014, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace SafetySharp.CSharp.Transformation
{
	using System;
	using System.Collections.Generic;
	using System.Linq;
	using Extensions;
	using Metamodel;
	using Metamodel.Declarations;
	using Microsoft.CodeAnalysis;
	using Microsoft.CodeAnalysis.CSharp.Syntax;
	using Utilities;

	/// <summary>
	///     Provides a mapping between C# symbols and <see cref="IMetamodelReference" />s.
	/// </summary>
	internal class SymbolMap
	{
		/// <summary>
		///     Maps a C# symbol by its name to a metamodel reference.
		/// </summary>
		private readonly Dictionary<string, IMetamodelReference> _symbolMap = new Dictionary<string, IMetamodelReference>();

		/// <summary>
		///     Initializes a new instance of the <see cref="SymbolMap" /> type.
		/// </summary>
		internal SymbolMap(Compilation compilation)
		{
			Argument.NotNull(compilation, () => compilation);

			foreach (var syntaxTree in compilation.SyntaxTrees)
				ResolveSymbols(compilation.GetSemanticModel(syntaxTree));
		}

		/// <summary>
		///     Gets a typed reference to the C# <paramref name="symbol" /> representing a component.
		/// </summary>
		/// <param name="symbol">The C# symbol the reference should be returned for.</param>
		internal IMetamodelReference<ComponentDeclaration> GetComponentReference(ITypeSymbol symbol)
		{
			Argument.NotNull(symbol, () => symbol);
			Argument.Satisfies(symbol.TypeKind == TypeKind.Class, () => symbol, "Expected a type symbol for a class.");

			return GetReference<ComponentDeclaration>(symbol);
		}

		/// <summary>
		///     Gets a typed reference to the C# <paramref name="symbol" /> representing an interface.
		/// </summary>
		/// <param name="symbol">The C# symbol the reference should be returned for.</param>
		internal IMetamodelReference<InterfaceDeclaration> GetInterfaceReference(ITypeSymbol symbol)
		{
			Argument.NotNull(symbol, () => symbol);
			Argument.Satisfies(symbol.TypeKind == TypeKind.Interface, () => symbol, "Expected a type symbol for an interface.");

			return GetReference<InterfaceDeclaration>(symbol);
		}

		/// <summary>
		///     Gets a typed reference to the C# <paramref name="symbol" /> representing a method.
		/// </summary>
		/// <param name="symbol">The C# symbol the reference should be returned for.</param>
		internal IMetamodelReference<MethodDeclaration> GetMethodReference(IMethodSymbol symbol)
		{
			Argument.NotNull(symbol, () => symbol);
			return GetReference<MethodDeclaration>(symbol);
		}

		/// <summary>
		///     Gets a typed reference to the C# <paramref name="symbol" /> representing a field.
		/// </summary>
		/// <param name="symbol">The C# symbol the reference should be returned for.</param>
		internal IMetamodelReference<FieldDeclaration> GetFieldReference(IFieldSymbol symbol)
		{
			Argument.NotNull(symbol, () => symbol);
			return GetReference<FieldDeclaration>(symbol);
		}

		/// <summary>
		///     Gets a typed reference to the C# <paramref name="symbol" />.
		/// </summary>
		/// <param name="symbol">The C# symbol the reference should be returned for.</param>
		private MetamodelReference<T> GetReference<T>(ISymbol symbol)
			where T : MetamodelElement
		{
			IMetamodelReference reference;
			if (!_symbolMap.TryGetValue(symbol.GetFullName(), out reference))
				throw new InvalidOperationException("The given C# symbol is unknown.");

			Assert.OfType<MetamodelReference<T>>(reference, "Expected a metamodel reference of type '{0}' but found '{1}'.",
												 typeof(MetamodelReference<T>).FullName, reference.GetType().FullName);

			return (MetamodelReference<T>)reference;
		}

		/// <summary>
		///     Gets a value indicating whether <paramref name="symbol" /> is mapped.
		/// </summary>
		/// <param name="symbol">The symbol that should be checked.</param>
		internal bool IsMapped(ISymbol symbol)
		{
			Argument.NotNull(symbol, () => symbol);
			return _symbolMap.ContainsKey(symbol.GetFullName());
		}

		/// <summary>
		///     Resolves all relevant symbols found within the <paramref name="semanticModel" />.
		/// </summary>
		/// <param name="semanticModel">The semantic model for the syntax tree that should be used to resolve the C# symbols.</param>
		private void ResolveSymbols(SemanticModel semanticModel)
		{
			foreach (var component in semanticModel.GetDeclaredComponents())
			{
				AddReference<ComponentDeclaration>(semanticModel, component);

				foreach (var method in component.DescendantNodes<MethodDeclarationSyntax>())
					AddReference<MethodDeclaration>(semanticModel, method);

				var fields = component
					.DescendantNodes<FieldDeclarationSyntax>()
					.Where(fieldDeclaration => !fieldDeclaration.IsComponentField(semanticModel))
					.SelectMany(fieldDeclaration => fieldDeclaration.Declaration.Variables);

				foreach (var field in fields)
					AddReference<FieldDeclaration>(semanticModel, field);
			}

			foreach (var componentInterface in semanticModel.GetDeclaredComponentInterfaces())
				AddReference<InterfaceDeclaration>(semanticModel, componentInterface);
		}

		/// <summary>
		///     Adds a <see cref="IMetamodelReference" /> instance for <paramref name="syntaxNode" /> to the symbol map.
		/// </summary>
		/// <typeparam name="T">The type of the metamodel element corresponding to the resolved symbol.</typeparam>
		/// <param name="semanticModel">The semantic model that should be used to resolve the C# symbol.</param>
		/// <param name="syntaxNode">The C# syntax node that should be resolved.</param>
		private void AddReference<T>(SemanticModel semanticModel, SyntaxNode syntaxNode)
			where T : MetamodelElement
		{
			var symbol = semanticModel.GetDeclaredSymbol(syntaxNode);
			Assert.NotNull(symbol, "The semantic model could not find a symbol for '{0}' '{1}'.", syntaxNode.GetType().FullName, syntaxNode);

			_symbolMap.Add(symbol.GetFullName(), new MetamodelReference<T>());
		}
	}
}

File: Source/SafetySharp/CSharp/Transformation/SymbolMap.fs
// The MIT License (MIT)
// 
// Copyright (c) 2014, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace SafetySharp.CSharp

open System.Collections.Generic
open System.Collections.Immutable
open System.Runtime.CompilerServices

open SafetySharp.Utilities
open SafetySharp.Metamodel
open SafetySharp.Modeling
open SafetySharp.CSharp.Roslyn

open Microsoft.CodeAnalysis
open Microsoft.CodeAnalysis.CSharp
open Microsoft.CodeAnalysis.CSharp.Syntax

/// Constructs symbols for the given component types and creates a mapping between the original C# symbols and the
/// created metamodel symbols.
type internal SymbolMap (semanticModel : SemanticModel, componentTypes : string list) =
    do Requires.NotNull semanticModel "semanticModel"
    do Requires.ArgumentSatisfies (componentTypes |> List.length > 0) "componentTypes" "At least one component type must be provided."

    // An equality comparer for method symbols that implements reference equality
    let comparer = { 
        new IEqualityComparer<MethodSymbol> with
            member this.Equals (symbol1, symbol2) = 
                obj.ReferenceEquals(symbol1, symbol2)
            member this.GetHashCode (symbol) =
                RuntimeHelpers.GetHashCode(symbol)
    }

    // We're using the builder pattern to initialize the dictionaries and the component list
    let componentListBuilder = ImmutableList.CreateBuilder<ComponentSymbol> ()
    let componentMapBuilder = ImmutableDictionary.CreateBuilder<INamedTypeSymbol, ComponentSymbol> ()
    let fieldMapBuilder = ImmutableDictionary.CreateBuilder<IFieldSymbol, FieldSymbol> ()
    let subComponentMapBuilder = ImmutableDictionary.CreateBuilder<IFieldSymbol, SubcomponentSymbol> ()
    let methodMapBuilder = ImmutableDictionary.CreateBuilder<IMethodSymbol, MethodSymbol> ()
    let methodMapBackBuilder = ImmutableDictionary.CreateBuilder<MethodSymbol, IMethodSymbol> (comparer)

    // Converts a C# type symbol to one of the supported metamodel type symbols
    let toTypeSymbol (csharpSymbol : ITypeSymbol) =
        match csharpSymbol.SpecialType with
        | SpecialType.None -> sprintf "Type '%A' is not yet supported." csharpSymbol.SpecialType |> invalidOp
        | SpecialType.System_Boolean -> TypeSymbol.Boolean
        | SpecialType.System_Int32 -> TypeSymbol.Integer
        | SpecialType.System_Decimal -> TypeSymbol.Decimal
        | _ -> sprintf "Unsupported C# type: '%A'." csharpSymbol.SpecialType |> invalidOp

    // Create the symbols for all components
    do for componentType in componentTypes |> Seq.distinct do
        let csharpComponent = semanticModel.GetTypeSymbol componentType
        if not <| csharpComponent.IsDerivedFromComponent semanticModel then
            sprintf "Type '%s' is not derived from '%s'." componentType typeof<Component>.FullName |> invalidOp

        let fields = csharpComponent.GetMembers().OfType<IFieldSymbol>()
        let methods = csharpComponent.GetMembers().OfType<IMethodSymbol>()
        let assemblyName = csharpComponent.ContainingAssembly.Identity.Name
        let componentName = csharpComponent.ToDisplayString SymbolDisplayFormat.MinimallyQualifiedFormat

        let componentSymbol = {
            // We're encoding the assembly and all parent namespaces in the component name to ensure the uniqueness of the name.
            Name = sprintf "%s::%s" assemblyName componentName

            // Creates the symbols and optional mapping information for the Update method of the component.
            UpdateMethod = 
                let updateMethods = methods |> Seq.filter (fun method' -> method'.IsUpdateMethod semanticModel)
                let updateMethodCount = updateMethods |> Seq.length
                let methodSymbol = { Name = "Update"; ReturnType = None; Parameters = [] }

                if updateMethodCount > 1 then 
                    sprintf "Component of type '%A' defines more than one Update() method." componentType |> invalidOp
                else if updateMethodCount = 1 then
                    let updateMethod = updateMethods |> Seq.head
                    methodMapBuilder.Add (updateMethod, methodSymbol)
                    methodMapBackBuilder.Add (methodSymbol, updateMethod)
                methodSymbol

            // Create the symbols and mapping information for all methods of the component. We'll also build up a 
            // dictionary that allows us to retrieve the original C# method symbol again.
            Methods = 
                [
                    let methods = methods |> Seq.filter (fun method' -> not <| method'.IsUpdateMethod semanticModel && method'.MethodKind = MethodKind.Ordinary)
                    for csharpMethod in methods ->
                        let methodSymbol = { Name = csharpMethod.Name; ReturnType = None; Parameters = []}
                        methodMapBuilder.Add (csharpMethod, methodSymbol)
                        methodMapBackBuilder.Add (methodSymbol, csharpMethod)
                        methodSymbol
                ]

            // Creates the symbols and mapping information for all fields of the component.
            Fields = 
                [
                    let fields = fields |> Seq.filter (fun field -> not <| field.IsSubcomponentField semanticModel)
                    for csharpField in fields -> 
                        let fieldSymbol = { FieldSymbol.Name = csharpField.Name; Type = toTypeSymbol csharpField.Type }
                        fieldMapBuilder.Add (csharpField, fieldSymbol)
                        fieldSymbol
                ]

            // Creates the symbols and mapping information for all subcomponents of the component.
            Subcomponents = 
                [
                    let fields = fields |> Seq.filter (fun field -> field.IsSubcomponentField semanticModel)
                    for csharpField in fields -> 
                        let subComponentSymbol = { SubcomponentSymbol.Name = csharpField.Name }
                        subComponentMapBuilder.Add (csharpField, subComponentSymbol)
                        subComponentSymbol
                ]
        }

        componentListBuilder.Add componentSymbol
        componentMapBuilder.Add (csharpComponent, componentSymbol)

    // Create the dictionaries and the component list that we'll later use to resolve C# symbols to metamodel symbols.
    let componentMap = componentMapBuilder.ToImmutable ()
    let componentList = componentListBuilder |> List.ofSeq
    let fieldMap = fieldMapBuilder.ToImmutable ()
    let subComponentMap = subComponentMapBuilder.ToImmutable ()
    let methodMap = methodMapBuilder.ToImmutable ()
    let methodMapBack = methodMapBackBuilder.ToImmutable ()

    /// Resolves the <see cref="ComponentSymbol"/> corresponding to the given C# component symbol.
    member this.Resolve (componentSymbol : INamedTypeSymbol) =
        Requires.NotNull componentSymbol "componentSymbol"
        Requires.That (componentMap.ContainsKey componentSymbol) "The given C# component symbol is unknown."
        componentMap.[componentSymbol]

    /// Resolves the <see cref="FieldSymbol"/> corresponding to the given C# field symbol.
    member this.ResolveField (fieldSymbol : IFieldSymbol) =
        Requires.NotNull fieldSymbol "fieldSymbol"
        Requires.That (fieldMap.ContainsKey fieldSymbol) "The given C# field symbol is unknown."
        fieldMap.[fieldSymbol]

    /// Resolves the <see cref="SubcomponentSymbol"/> corresponding to the given C# subcomponent symbol.
    member this.ResolveSubcomponent (subcomponentSymbol : IFieldSymbol) =
        Requires.NotNull subcomponentSymbol "subcomponentSymbol"
        Requires.That (fieldMap.ContainsKey subcomponentSymbol) "The given C# subcomponent symbol is unknown."
        subComponentMap.[subcomponentSymbol]

    /// Resolves the <see cref="MethodSymbol"/> corresponding to the given C# method symbol.
    member this.Resolve (methodSymbol : IMethodSymbol) =
        Requires.NotNull methodSymbol "methodSymbol"
        Requires.That (methodMap.ContainsKey methodSymbol) "The given C# method symbol is unknown."
        methodMap.[methodSymbol]

    /// Resolves the C# method symbol corresponding to the given metamodel <see cref="MethodSymbol"/>.
    member this.Resolve (methodSymbol : MethodSymbol) =
        Requires.That (methodMapBack.ContainsKey methodSymbol) "The given method symbol is unknown."
        methodMapBack.[methodSymbol]

    /// Gets a list of all component symbols.
    member this.Components = componentList